hawkular-alerts-bus/src/main/java/org/hawkular/alerts/bus/init/CacheManager.java

```diff
@@ -46,98 +51,236 @@
+                } else {
+                    updatedSubscribers = new HashSet<>();
+                    updatedSubscribers.add(ALERTING);
+                    publishCache.put(metricId, updatedSubscribers);
+                }
+            */
+        }
+    }
+
+    private void unpublish(DataIdKey dataIdKey) {
+        MetricId metricId = getMetricId(dataIdKey);
+        if (metricId != null && publishCache != null) {
+            /*
+                This logic assumes that alerting is the only writer for the shared publishCache
+             */
+            publishCache.remove(metricId);
+            /*
+                This alternative logic assumes that more writers can add/remove entries into the shared publishCache.
+
+                if (publishCache.containsKey(metricId)) {
+                    Set<String> updatedSubscribers;
+                    Set<String> subscribers = (Set<String>) publishCache.get(metricId);
+                    if (subscribers.contains(ALERTING)) {
+                        updatedSubscribers = new HashSet<>(subscribers);
+                        updatedSubscribers.remove(ALERTING);
+                        if (updatedSubscribers.isEmpty()) {
+                            publishCache.remove(metricId);
+                        } else {
+                            publishCache.put(metricId, updatedSubscribers);
+                        }
+                    }
+                }
+            */
+        }
+    }
+
+    private MetricId getMetricId(DataIdKey dataIdKey) {
+        String tenantId = dataIdKey.getTenantId();
+        String dataId = dataIdKey.getDataId();
+        String metricId;
+        MetricType type;
+        if (dataId.startsWith(AVAILABILITY)) {
+            type = MetricType.AVAILABILITY;
+            metricId = dataId.substring(AVAILABILITY.length() + 1);
+        } else if (dataId.startsWith(GAUGE)) {
+            type = MetricType.GAUGE;
+            metricId = dataId.substring(GAUGE.length() + 1);
+        } else if (dataId.startsWith(COUNTER)) {
+            type = MetricType.COUNTER;
+            metricId = dataId.substring(COUNTER.length() + 1);
+        } else if (dataId.startsWith(COUNTER_RATE)) {
+            type = MetricType.COUNTER_RATE;
+            metricId = dataId.substring(COUNTER_RATE.length() + 1);
+        } else if (dataId.startsWith(STRING)) {
+            type = MetricType.STRING;
+            metricId = dataId.substring(STRING.length() + 1);
+        } else if (dataId.startsWith(GAUGE_RATE)) {
+            type = MetricType.GAUGE_RATE;
+            metricId = dataId.substring(GAUGE_RATE.length() + 1);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("DataId " + dataId + " doesn't have a valid metrics type.");
+            }
+            return null;
+        }
+        return new MetricId(tenantId, type, metricId);
+    }
+
+    public static class DataIdKey {
+        private String tenantId;
+        private String dataId;
+
```

Comment:
perhaps just store the metricId in here as another field and strip out the prefix in the constructor.